### ISSUE TITLE
Cannot find libpq.5.dylib after updating to postgres 14.5

### ISSUE THREAD

**Issue by mattpetersen** [psycopg/psycopg2#1490]
# tl; dr

Homebrew PostgreSQL installation moved the location of the libpq.

Reinstalling psycopg2 should fix the problem.

---

I'm using the latest `postgresql@14`, installed by `brew`.

I was getting the following error today

```text
Traceback (most recent call last):
  File "/Users/matt/servertest/diff/snapshot.py", line 12, in <module>
    from app import app
  File "/Users/matt/server-core/app/__init__.py", line 9, in <module>
    import psycopg2.extensions
  File "/Users/matt/.pyenv/versions/nammu-backend-3.10.3/lib/python3.10/site-packages/psycopg2/__init__.py", line 51, in <module>
    from psycopg2._psycopg import (                     # noqa
ImportError: dlopen(/Users/matt/.pyenv/versions/nammu-backend-3.10.3/lib/python3.10/site-packages/psycopg2/_psycopg.cpython-310-darwin.so, 0x0002): Library not loaded: '/usr/local/opt/postgresql/lib/libpq.5.dylib'
  Referenced from: '/Users/matt/.pyenv/versions/3.10.3/envs/nammu-backend-3.10.3/lib/python3.10/site-packages/psycopg2/_psycopg.cpython-310-darwin.so'
  Reason: tried: '/usr/local/opt/postgresql/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file), '/usr/local/Cellar/postgresql@14/14.5_1/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file)
```

It started after running

```bash
brew update && brew upgrade
```

The reason is that the `libpq.5.dylib` file is expected in

```text
/usr/local/Cellar/postgresql@14/14.5_1/lib/libpq.5.dylib
```

But is actually in

```text
/usr/local/Cellar/postgresql@14/14.5_1/lib/postgresql@14/libpq.5.dylib
```

The solution was to reinstall `psycopg2`

```bash
pip reinstall pyscopg2
```

There was not a new version of `psycopg2` available. I was already using `psycopg2` version 2.9.3.

Rather, it might be that, when installing, `psycopg2` looks for and caches the path to the `libpq.5.dylib` file, which has changed in the latest version of `postgres`. Maybe someone here can confirm this?

Before today, I was using `postgres` 14.4 without issue. The problematic update was to `postgres` 14.5, which was released on August 11th, 2022. It was available as an update via `brew` sometime between August 24th and 27th, 2022.

Sorry to post a resolved issue, but I thought it might help someone.

**Comment by syphar** [comment]
@mattpetersen thanks for the hint! I ran into the same issue today

**Comment by jaklan** [comment]
As I just mentioned in https://github.com/Homebrew/homebrew-core/issues/109164:
> I found a reason why reinstalling `dbt` wasn't enough:
> https://github.com/dbt-labs/dbt-core/blob/1df713fee986741289568f553befdd4e38a51122/plugins/postgres/setup.py#L32-L40
> `dbt` by default uses `psycopg2-binary`, not `psycopg2`. And it seems reinstalling `psycopg2-binary` doesn't solve the issue, but additionally installing pure `psycopg2` - does.
So maybe we should re-open the issue if `psycopg2-binary` is still affected?

**Comment by proxseas** [comment]
For the record, re-installing "psycopg2-binary" also works (if you're using that instead of just "psycopg2").

**Comment by jaklan** [comment]
> For the record, re-installing "psycopg2-binary" also works (if you're using that instead of just "psycopg2").

That's interesting, because I have verified that twice in a clear environment where I installed dbt-redshift and reinstalling the binary package didn't help at all, I had to play with the source package.

**Comment by syphar** [comment]
> > For the record, re-installing "psycopg2-binary" also works (if you're using that instead of just "psycopg2").
> 
> That's interesting, because I have verified that twice in a clear environment where I installed dbt-redshift and reinstalling the binary package didn't help at all, I had to play with the source package.

you also have to clear the pip / wheel cache locally. 

As I understood the issue this can only happen when the wheel was locally built, then cached. A reinstall won't rebuild the local wheel but only pull it from the cache

**Comment by proxseas** [comment]
> > For the record, re-installing "psycopg2-binary" also works (if you're using that instead of just "psycopg2").
> 
> That's interesting, because I have verified that twice in a clear environment where I installed dbt-redshift and reinstalling the binary package didn't help at all, I had to play with the source package.

Hmm, well then I guess I could give some context on what my set up is like. My system is an M1 Macbook Air. Yesterday I started experiencing issues with my Postgres installation, which was installed through Homebrew. This was related to upgrading to the latest version of postgres - v14.5, I believe. Some of the paths changed - and this broke many things. I tried fixing it in various ways, including a full re-install of Postgres through Homebrew, until I decided to try out "Postgres app" ([link](https://postgresapp.com/)). This worked immediately and made my Db accessible and usable again through PSQL and pgAdmin (I had backups available, so I just used 'pg_restore' to get my data restored). However, I still wasn't able to interact with it using Python.

Relatedly, the way I was using Postgres in Python was through PipEnv. For the longest time, I had 'psycopg2-binary' and just 'pscopg' (which is version 3.x) installed in that same PipEnv virtual env. I think having the latter package didn't really affect anything. I had even forgotten that it was around.

In any case, these are the steps I took today to fix the issue I had with psycopg:
* Removed 'psycopg' (v3.x): 'pipenv uninstall psycopg' (this is probably not relevant)
* 'pipenv uninstall psycopg2-binary'
* 'pipenv install psycopg2-binary'

I didn't do any cache clearing - it worked immediately.

**Comment by dvarrazzo** [comment]
> My system is an M1 Macbook Air

psycopg2-binary doesn't support M1 hardware. You are just installing psycopg2 from source and you are depending on your local libpq

**Comment by proxseas** [comment]
> > My system is an M1 Macbook Air
> 
> psycopg2-binary doesn't support M1 hardware. You are just installing psycopg2 from source and you are depending on your local libpq

That could be the case (behind the scenes). But I just verified that the only package installed right now (with my current working setup) is "psycopg2-binary" version 2.9.3.

**Comment by jaklan** [comment]
@proxseas it seems to be really weird in general, because in my case ignoring cache helps, but... only once: https://github.com/psycopg/psycopg2/issues/1492#issuecomment-1233781034

**Comment by dvarrazzo** [comment]
As far as I can see,

**Comment by proxseas** [comment]
Quick update: I had a similar problem in another environment where I also use psycopg2. In this one, I had both 'psycopg2-binary' and 'psycopg2' installed (as shown by 'pipenv graph'). These are the steps I took to fix it:
* pipenv uninstall psycopg2
* pipenv uninstall psycopg2-binary
* pipenv lock --clear
* pipenv --clear
* pip cache purge
* pipenv install psycopg2-binary

**Comment by adriaant** [comment]
The problem is that the latest postgresql install script creates another `postgresql@14` directory. Instead of:
homebrew/Cellar/postgresql@14/14.5_1/lib/<libs>
you get
homebrew/Cellar/postgresql@14/14.5_1/lib/postgresql@14/<libs>

**Comment by dvarrazzo** [comment]
@mattpetersen thank you for the report. The Homebrew maintainers know of the problem. No idea when will be fixed, however, reinstalling psycopg2, should fix the problem for the people affected by it.

**Comment by hf-krechan** [comment]
Did I understand it correct, that you mean with reinstalling `psycopg2` that I should run the following commands in my python virtual environment
```bash
pip uninstall psycopg2
pip install psycopg2
```

Because this does not solve the problem for me.
I am using a MacBook Pro with a M1 Pro processor.
There is still the same error message:
```
Reason: tried: '/opt/homebrew/opt/postgresql/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file), '/opt/homebrew/Cellar/postgresql@14/14.5_2/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file)
```

**Comment by jaklan** [comment]
@hf-krechan https://github.com/psycopg/psycopg2/issues/1492#issuecomment-1233930135

**Comment by proxseas** [comment]
> Did I understand it correct, that you mean with reinstalling `psycopg2` that I should run the following commands in my python virtual environment
> 
> ```shell
> pip uninstall psycopg2
> pip install psycopg2
> ```
> 
> Because this does not solve the problem for me. I am using a MacBook Pro with a M1 Pro processor. There is still the same error message:
> 
> ```
> Reason: tried: '/opt/homebrew/opt/postgresql/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file), '/opt/homebrew/Cellar/postgresql@14/14.5_2/lib/libpq.5.dylib' (no such file), '/usr/local/lib/libpq.5.dylib' (no such file), '/usr/lib/libpq.5.dylib' (no such file)
> ```

What I did before reinstalling psycopg2 was:

1. Uninstall postgresql server entirely from Homebrew. I removed some directories manually.
2. Install postgres server through https://postgresapp.com/. It's one of the officially recommended installation methods and it has worked very well for me so far.

**Comment by hf-krechan** [comment]
I am a little bit confused. For one of our python based backend systems we use a Postgres Docker container for local development.
Therefor I do not need a postgres server running on my local machine.
Or comes the postgres server with the binary to connect to a postgres server which psycopg2 uses?

**Comment by hf-krechan** [comment]
> @hf-krechan [#1492 (comment)](https://github.com/psycopg/psycopg2/issues/1492#issuecomment-1233930135)

In the [psycopg2 documentation](https://www.psycopg.org/docs/install.html#psycopg-vs-psycopg-binary) they mention the following
> For production use you are advised to use the source distribution.

So you should not use the `psycopg2-binary` for production.
I would like to avoid to write the `psycopg2-binary` dependency into the `requirements.txt` cause it is also used in our CI/CD pipeline.

So I would be happy to still use homebrew to install postgres, but I need a working psycopg2 ^^

**Comment by jaklan** [comment]
@hf-krechan well, tell that to the `dbt` guys - they use `psycopg2-binary` by default. 

Anyway - the same flow with clearing cache as I described in the linked commend works for the source-based `psycopg2`.

**Comment by hf-krechan** [comment]
Uh nice this worked :)
Thank you very much.
For the people who came from their favourite search engine, these are my steps I made in the virtual environment:

```bash
# uninstall psycopg2
pip uninstall psycopg2
# clear cache of psycopg2
pip cache remove psycopg2
# install it again
pip install psycopg2
```

**Comment by Jacechrome-options** [comment]
> The problem is that the latest postgresql install script creates another `postgresql@14` directory. Instead of: homebrew/Cellar/postgresql@14/14.5_1/lib/... you get homebrew/Cellar/postgresql@14/14.5_1/lib/postgresql@14/...





https://urlencodeunpopular.github.io/brydon224/
